### PR TITLE
Update recommendation on `dataloader_idx`

### LINKED ISSUE #10315: Always pass dataloader_idx to step functions.
## Proposed refactoring or deprecation

Pass `dataloader_idx` to `[train,test,val,inference]_step` regardless of the number dataloaders.

### Motivation

I often use the same lightning modules in different projects, ie. I have one `Classifier` task that I use when I need to train classifiers. Now `Trainer` passes either 3 or 4 arguments to `step` functions depending on whether I have one or multiple loaders.

```python
# Fails with multiple dataloaders.
def training_step(self, batch, batch_idx):
    return  self(batch).sum()

# Fails if with only one DataLoader.
def training_step(self, batch, batch_idx, dataloader_idx):
    return  self(batch).sum()
```

A quick fix that could be used (ie. I'm using) is to define the training step as below but it's not pretty... :D

```python
def training_step(self, batch, batch_idx, *args):
    dataloader_idx = 0 if len(args) == 0 else args[0]
```

I think passing the dataloader_idx anyways to step functions would simplify `LightningModule` as the module itself wouldn't have to be changed depending on the number of dataloaders.

### Pitch

Always pass `dataloader_idx` to `step` functions.

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #10315

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]
king of grep

**Review by tchaton — APPROVED** [review]
LGMT !

**Review by ethanwharris — APPROVED** [review]

**Review by Borda — APPROVED** [review]
